### PR TITLE
Avoid mishandling retain_n of immortal objects where n >= 3

### PR DESCRIPTION
**Pull request by Catfish-Man**
Fixes rdar://problem/59547400

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please test

File: stdlib/public/SwiftShims/RefCount.h [hunk 1]
```diff
@@ -1248,7 +1252,15 @@ class RefCounts {
   // Return weak reference count.
   // Note that this is not equal to the number of outstanding weak pointers.
   uint32_t getWeakCount() const;
-  
+
+  // DO NOT TOUCH.
+  // This exists for the benefits of the Refcounting.cpp tests. Do not use it
+  // elsewhere.
+  auto getBitsValue()
+    -> decltype(refCounts.load(std::memory_order_relaxed).getBitsValue()) {
```
Comment (CodaFi) [inline stdlib/public/SwiftShims/RefCount.h]: I believe `decltype(auto)` will get you this behavior nowadays.

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
hm. Interesting… I wonder what I'm missing here.

File: stdlib/public/SwiftShims/RefCount.h (hunk 1 above)
Comment (mikeash) [reply stdlib/public/SwiftShims/RefCount.h]: I thought that you needed C++14 for return type inference on a method.

Comment (CodaFi) [reply stdlib/public/SwiftShims/RefCount.h]: https://github.com/apple/swift/commit/8246864260eb6ee8e7ee62f2c60e58a9d428ff1e

Comment (mikeash) [reply stdlib/public/SwiftShims/RefCount.h]: I'm behind the times! Thanks.

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Pull request merged**